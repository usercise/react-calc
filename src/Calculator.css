/* Change size of rem for small screens so we fit all the way down to 150px width */
@media (max-width: 250px) {
  html {
    font-size: 8px;
  }
}

.Calculator {
  max-width: 30rem;
  padding: 2rem;
}

.Calculator--large {
  max-width: none;
}

.Calculator-result {
  background-color: rgba(161, 161, 197, 0.6);
  text-align: right;
  padding: 0.5rem 1rem;
  margin-bottom: 0;
  overflow: hidden;
  text-overflow: ellipsis;
}

.Calculator-inputs {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  grid-gap: 1rem;
  background-color: rgba(17, 17, 22, 0.9);
  color: #fff;
  padding: 1rem;
}
