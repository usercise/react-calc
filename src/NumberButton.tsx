import React from 'react';
import Button from './Button';

interface NumberButtonProps {
  number: number;
  displayValue: string;
  isEditing: boolean;
  updateValue: (newValue: string) => void;
}

const NumberButton: React.SFC<NumberButtonProps> = props => {
  const strNumber = `${props.number}`;
  const addValue = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    const newValue =
      !props.isEditing || props.displayValue === '0'
        ? strNumber
        : `${props.displayValue}${props.number}`;
    props.updateValue(newValue);
  };
  return (
    <Button title={strNumber} buttonValue={props.number} callback={addValue} />
  );
};
export default NumberButton;
