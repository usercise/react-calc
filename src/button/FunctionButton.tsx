import React from 'react';
import Button from './Button';

interface FunctionButtonProps {
  call: string;
  displayValue: string;
  modifier?: string;
  functionButtonUpdateValue: (newFunction: string, isEditing: boolean) => void;
}

const FunctionButton: React.SFC<FunctionButtonProps> = props => {
  const buttonCall = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    props.functionButtonUpdateValue(props.call, false);
  };
  return (
    <Button
      modifier={props.modifier}
      title={props.call}
      callback={buttonCall}
    />
  );
};
export default FunctionButton;
