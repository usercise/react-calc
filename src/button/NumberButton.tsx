import React from 'react';
import Button from './Button';

interface NumberButtonProps {
  number: number;
  displayValue: string;
  isEditing: boolean;
  updateInput: (newValue: string) => void;
}

const NumberButton: React.SFC<NumberButtonProps> = props => {
  const strNumber = props.number.toString();
  const addValue = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    const newValue =
      !props.isEditing || props.displayValue === '0'
        ? strNumber
        : `${props.displayValue}${props.number}`;
    props.updateInput(newValue);
  };
  return (
    <Button title={strNumber} buttonValue={props.number} callback={addValue} />
  );
};
export default NumberButton;
