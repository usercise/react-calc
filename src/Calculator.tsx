import React, { Component } from 'react';
import NumberButton from './button/NumberButton';
import AdditionButton from './button/AdditionButton';
import './Calculator.css';

interface CalculatorProps {
  isLargeSize: boolean;
}

interface CalculatorState {
  entryValue: string;
  stateValue: number;
  isEditing: boolean;
}

class Calculator extends Component<CalculatorProps, CalculatorState> {
  constructor(props: CalculatorProps) {
    super(props);
    this.state = {
      entryValue: '0',
      stateValue: 0,
      isEditing: false
    };
  }

  updateInput(newValue: string) {
    this.setState({ entryValue: newValue, isEditing: true });
  }

  updateValue(newValue: number, isEditing: boolean) {
    this.setState({
      stateValue: newValue,
      entryValue: newValue.toString(),
      isEditing
    });
  }

  render() {
    return (
      <div
        className={`Calculator${
          this.props.isLargeSize ? ' Calculator--large' : ''
        }`}
      >
        <h2 className="Calculator-result">{this.state.entryValue}</h2>
        <div className="Calculator-inputs">
          <button
            onClick={e => this.setState({ entryValue: '0', isEditing: false })}
          >
            C
          </button>
          <button>/</button>
          <button>*</button>
          <button>-</button>
          <NumberButton
            number={7}
            isEditing={this.state.isEditing}
            displayValue={this.state.entryValue}
            updateInput={this.updateInput.bind(this)}
          />
          <NumberButton
            number={8}
            isEditing={this.state.isEditing}
            displayValue={this.state.entryValue}
            updateInput={this.updateInput.bind(this)}
          />
          <NumberButton
            number={9}
            isEditing={this.state.isEditing}
            displayValue={this.state.entryValue}
            updateInput={this.updateInput.bind(this)}
          />
          <AdditionButton
            updateValue={this.updateValue.bind(this)}
            number={this.state.stateValue}
            displayValue={this.state.entryValue}
          />

          {/* <button className="Calculator-plus">+</button> */}
          <NumberButton
            number={4}
            isEditing={this.state.isEditing}
            displayValue={this.state.entryValue}
            updateInput={this.updateInput.bind(this)}
          />
          <NumberButton
            number={5}
            isEditing={this.state.isEditing}
            displayValue={this.state.entryValue}
            updateInput={this.updateInput.bind(this)}
          />
          <NumberButton
            number={6}
            isEditing={this.state.isEditing}
            displayValue={this.state.entryValue}
            updateInput={this.updateInput.bind(this)}
          />
          <button className="Calculator-equal">=</button>
          <NumberButton
            number={1}
            isEditing={this.state.isEditing}
            displayValue={this.state.entryValue}
            updateInput={this.updateInput.bind(this)}
          />
          <NumberButton
            number={2}
            isEditing={this.state.isEditing}
            displayValue={this.state.entryValue}
            updateInput={this.updateInput.bind(this)}
          />
          <NumberButton
            number={3}
            isEditing={this.state.isEditing}
            displayValue={this.state.entryValue}
            updateInput={this.updateInput.bind(this)}
          />
          <NumberButton
            number={0}
            isEditing={this.state.isEditing}
            displayValue={this.state.entryValue}
            updateInput={this.updateInput.bind(this)}
          />
          <button>.</button>
        </div>
      </div>
    );
  }
}

export default Calculator;
