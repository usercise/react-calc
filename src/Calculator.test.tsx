import React from 'react';
import { mount, shallow, render } from 'enzyme';
import NumberButton from './button/NumberButton';
import FunctionButton from './button/FunctionButton';
import Calculator from './Calculator';

describe('<Calculator />', () => {
  it('renders without crashing', () => {
    const component = shallow(<Calculator isLargeSize={false} />);
    expect(component.html()).toMatchSnapshot();
  });

  it('large class not to be set when isLargeSize is false', () => {
    const component = shallow(<Calculator isLargeSize={false} />);
    expect(component.hasClass('Calculator--large')).toBeFalsy();
  });

  it('large class to be set when isLargeSize is true', () => {
    const component = shallow(<Calculator isLargeSize={true} />);
    expect(component.hasClass('Calculator--large')).toBeTruthy();
  });

  it('display to show zero', () => {
    const component = shallow(<Calculator isLargeSize={false} />);
    expect(component.find('.Calculator-result').text()).toBe('0');
  });

  it('display to show seven after clicking first number button', () => {
    const component = mount(<Calculator isLargeSize={false} />);
    const numberSeven = component.find(NumberButton).first();
    numberSeven.simulate('click');
    expect(component.state('entryValue')).toBe('7');
    expect(component.find('.Calculator-result').text()).toBe('7');
  });

  it('display to show 1 after clicking 1 + 1', () => {
    const component = mount(<Calculator isLargeSize={false} />);
    const numberOne = component.find(NumberButton).at(6);
    const functionPlus = component.find(FunctionButton).at(5);
    numberOne.simulate('click');
    expect(component.state('entryValue')).toBe('1');
    functionPlus.simulate('click');
    expect(component.state('stateValue')).toBe(1);
    numberOne.simulate('click');
    expect(component.state('entryValue')).toBe('1');

    expect(component.find('.Calculator-result').text()).toBe('1');
  });

  it('display to show 2 after clicking 1 + 1 +', () => {
    const component = mount(<Calculator isLargeSize={false} />);
    const numberOne = component.find(NumberButton).at(6);
    const functionPlus = component.find(FunctionButton).at(5);
    numberOne.simulate('click');
    expect(component.state('entryValue')).toBe('1');
    functionPlus.simulate('click');
    expect(component.state('stateValue')).toBe(1);
    numberOne.simulate('click');
    expect(component.state('entryValue')).toBe('1');
    functionPlus.simulate('click');
    expect(component.state('stateValue')).toBe(2);

    expect(component.find('.Calculator-result').text()).toBe('2');
  });

  it('display to show 2 after clicking 1 + 1 =', () => {
    const component = mount(<Calculator isLargeSize={false} />);
    const numberOne = component.find(NumberButton).at(6);
    const functionPlus = component.find(FunctionButton).at(5);
    const functionEqual = component.find(FunctionButton).last();
    numberOne.simulate('click');
    expect(component.state('entryValue')).toBe('1');
    functionPlus.simulate('click');
    expect(component.state('stateValue')).toBe(1);
    numberOne.simulate('click');
    expect(component.state('entryValue')).toBe('1');
    functionEqual.simulate('click');
    expect(component.state('stateValue')).toBe(2);

    expect(component.find('.Calculator-result').text()).toBe('2');
  });
});
