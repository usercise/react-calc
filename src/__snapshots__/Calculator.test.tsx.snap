// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator /> renders without crashing 1`] = `"<div class=\\"Calculator\\"><h2 class=\\"Calculator-result\\">0</h2><div class=\\"Calculator-inputs\\"><button class=\\"Button\\" type=\\"button\\">C</button><button class=\\"Button\\" type=\\"button\\">/</button><button class=\\"Button\\" type=\\"button\\">*</button><button class=\\"Button\\" type=\\"button\\">-</button><button class=\\"Button\\" type=\\"button\\" value=\\"7\\">7</button><button class=\\"Button\\" type=\\"button\\" value=\\"8\\">8</button><button class=\\"Button\\" type=\\"button\\" value=\\"9\\">9</button><button class=\\"Button Button--plus\\" type=\\"button\\">+</button><button class=\\"Button\\" type=\\"button\\" value=\\"4\\">4</button><button class=\\"Button\\" type=\\"button\\" value=\\"5\\">5</button><button class=\\"Button\\" type=\\"button\\" value=\\"6\\">6</button><button class=\\"Button Button--equal\\" type=\\"button\\">=</button><button class=\\"Button\\" type=\\"button\\" value=\\"1\\">1</button><button class=\\"Button\\" type=\\"button\\" value=\\"2\\">2</button><button class=\\"Button\\" type=\\"button\\" value=\\"3\\">3</button><button class=\\"Button Button--zero\\" type=\\"button\\" value=\\"0\\">0</button><button class=\\"Button\\" type=\\"button\\">.</button></div></div>"`;
